### PR TITLE
#162949077 Enable User to Rate an Article

### PR DESCRIPTION
**Pull request by waracci**
### What does this PR do?
Allows a user to rate an article they have not authored.

### Description of Task to be completed?
- [x] Add tests
- [x] Create a rating app
- [x] Create a model for rating
- [x] Add rating serializers
- [x] Create rating views
- [x] Have the following endpoints working:
- [x] POST ```/api/v1/rate/slug/```
### How should this be manually tested?
1. Clone this repository on your local machine using 

    ```$```  ```git clone https://github.com/andela/ah-robotics.git```
2. Change your working directory to the project's directory using 
     
     ```$``` ```cd ah-robotics```
3. Change to the specific branch `ft-add-article-rating-162949077` using 

     ```$``` ```git checkout ft-add-article-rating-162949077```
4. Create a virtual environment using the following command: 
     ```$```  ```python3 -m virtualenv env```
5. Activate the virtual environment using 
```$``` ```source env/bin/activate```
6. Install all the requirements using 
     ```$``` ```pip install -r requirements.txt```

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
       ```$``` ```pg_ctl -D /usr/local/var/postgres start```
    
    2. Run psql interactive terminal
    
        ```$``` ```psql postgres```
    
    3. Create a new user:
    
        ```$``` ```CREATE USER sample_username WITH PASSWORD 'sample_password';```
    
    4. Grant privileges to the user:
    
        ```$``` ```ALTER USER sample_username CREATEDB;```
    
    5. Create a database:
    
       ```$``` ```CREATE DATABASE sample_database_name WITH OWNER sample_username;```
    
    6. Exit psql interactive with the following command: ```$``` ```\q```
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get cloudinary values, create an account with cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      ```$``` ```source .env```
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      ```$``` ```python manage.py makemigrations```
      
      (ii) Apply the migrations using: 
      
      ```$``` ```python manage.py migrate```

10. Run the following command to collect static files.

   ```$``` ```python manage.py collectstatic```
      
#### Testing the features

1. Run the server from the terminal using: 
    
  ```$```  ```python manage.py runserver```
2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL `http://127.0.0.1:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and then add a payload to the body using the following format:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

**Example** 
sample_username---> honululu, 
sample_email-->hanku@example.com, 
password-->test1234

5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Log in the user using the following endpoint `http://127.0.0.1:8000/api/v1/users/login/` and insert the following payload to the body section.
    
    ``` {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

**NB** Use details you created in preceding step ```4```
7. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
8. After logging in the screen should look as follows:
    
    
screenshot 2019-01-29 at 15 59 01

9. Copy the token and click on the **Headers** section. In the **Key** column, type in the word **Authorization**. And in its corresponding value in the **Value** column, write the word __token__, add a space and paste in the token that you copied from the login response. The **Headers** section should look as follows after doing so:
    
    image

10. To Create an article, insert the following URL: http://127.0.0.1:8000/api/v1/articles/. In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, post an article by using the following payload:
    
    ```
    {
        "article": {
            "title": "Sample Article",
            "description": "Sample Article Description",
            "body": "The sample article body"
        }
    }
    ```
    Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.

11. To get all the articles, use the URL http://127.0.0.1:8000/api/v1/articles/. Change the method to a **GET** and hit **Send**.

12. To get a specific article, Copy the slug and insert it at the end of the articles URL as follows:
 http://127.0.0.1:8000/api/v1/articles/slug/.  Change the method to a **GET** and hit **Send**. 
For a example, to get a specific article with **my-first-article** as the slug name, the URL would be as follows: http://127.0.0.1:8000/api/v1/articles/my-first-article/. 
    > **Note**: The slug is a field in the article response body

13. On the PARAM paste the URL: ```http://127.0.0.1:8000/api/v1/rate/slug/```

**NB** Where ```slug```---> ```sample-article```

```http://127.0.0.1:8000/api/v1/rate/sample-article/```

screenshot 2019-01-29 at 16 24 54

14. On the body pass the following payload.

```
{
	"user_rating": 2
}
```
**NB:** The rating can be a value between ```1``` and ```5```

15. Given that you are the user who created this particular post this is the screenshot you will encounter.

screenshot 2019-01-29 at 16 12 17

16. Open a new tab on POSTMAN are create a different user following steps ```2``` to ```9``` as before.

17. Then paste the URL  ```http://127.0.0.1:8000/api/v1/rate/sample-article/``` on the PARAM. Repeat step ```14```
You will get a successful rating message as depicted below.  

screenshot 2019-01-29 at 16 44 45

**If you had multiple users registered you will be able to test that the average rating is also working**.

#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#162949077](https://www.pivotaltracker.com/n/projects/2235171)

## Questions:
Questions and comments are welcome.

### CODE REVIEW

**Review by jomasim — COMMENTED** [review]
.DS_Store binary files should not pushed just check the branch and unstage the file.

**Comment by jomasim** [comment]
- use related story link instead of project link

**Review by johnwayodi — COMMENTED** [review]
Tested it and it works. However, the branch seems to have some conflicts, kindly look into that.

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @waracci and @Teatoller,
I have tested the functionality and it works as expected. However, fix the merge conflicts to update your branch.

File: authors/apps/rating/views.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+from django.shortcuts import get_object_or_404
+from rest_framework.generics import (
+    GenericAPIView
+)
+from .models import Article, Rating
```
Comment (sharkdevs) [inline authors/apps/rating/views.py:5]: separate your different types of imports

File: authors/apps/articles/serializers.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+from rest_framework import serializers
+from django.db.models import Avg
+
+from authors.apps.profiles.serializers import AuthorSerializer
+from authors.apps.profiles.models import UserProfile
+
```
Comment (sharkdevs) [inline authors/apps/articles/serializers.py:8]: let the local application imports be in the same block

**Review by sharkdevs — COMMENTED** [review]
Those were many files. you can consider the changes below. Good work.

**Review by johnwayodi — APPROVED** [review]

**Pull request merged**